includes/abstracts/abstract-wc-order.php

```diff
@@ -983,22 +983,47 @@ public function remove_coupon( $code ) {
 	 * @since  3.2.0
 	 */
 	protected function recalculate_coupons() {
-		$discounts = new WC_Discounts( $this );
-		$coupons   = $this->get_items( 'coupon' );
-
-		foreach ( $coupons as $coupon ) {
-			$coupon_object = new WC_Coupon( $coupon->get_code() );
-			$discounts->apply_coupon( $coupon_object );
-		}
-
-		$this->set_coupon_discount_amounts( $discounts );
-
 		// Reset line item totals.
 		foreach ( $this->get_items() as $item ) {
 			$item->set_total( $item->get_subtotal() );
 			$item->set_total_tax( $item->get_subtotal_tax() );
 		}
 
+		$discounts = new WC_Discounts( $this );
+
+		foreach ( $this->get_items( 'coupon' ) as $coupon_item ) {
+			$coupon_code   = $coupon_item->get_code();
+			$coupon_id     = wc_get_coupon_id_by_code( $coupon_code );
+			$coupon_object = false;
+
+			if ( $coupon_id ) {
+				$coupon_object = new WC_Coupon( $coupon_id );
```

Comment:
The constructor allows to populate `WC_Coupon` by supplying the coupon code. Looks like we can get rid of calling `wc_get_coupon_id_by_code` to avoid duplicating code.

On the other hand, we need to be careful that `$coupon_code` is string, otherwise it will fall into the id check.

Reply:
Something along the lines:
```
$coupon_object = new WC_Coupon( (string) $coupon_code );

if ( ! $coupon_object->get_id() && $original_coupon_data = $coupon_item->get_meta( 'coupon_data', true ) ) {
```

Reply:
It's here so we can construct a new object should it not match a real coupon. Otherwise the _data_ filter kicks in and plugins like P&R try to make a manual coupon again.